waiter/object_delete.go

```diff
@@ -0,0 +1,84 @@
+package waiter
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"time"
+
+	"github.com/nspcc-dev/neofs-sdk-go/client"
+	apistatus "github.com/nspcc-dev/neofs-sdk-go/client/status"
+	cid "github.com/nspcc-dev/neofs-sdk-go/container/id"
+	oid "github.com/nspcc-dev/neofs-sdk-go/object/id"
+	"github.com/nspcc-dev/neofs-sdk-go/session"
+)
+
+// ObjectDeleteExecutor represents requirements to async object delete operation.
+type ObjectDeleteExecutor interface {
+	ObjectDelete(ctx context.Context, containerID cid.ID, objectID oid.ID, prm client.PrmObjectDelete) (oid.ID, error)
+	ObjectHead(ctx context.Context, containerID cid.ID, objectID oid.ID, prm client.PrmObjectHead) (*client.ResObjectHead, error)
+}
+
+// ObjectDeleteWaiter implements sync logic to object delete operation.
+type ObjectDeleteWaiter struct {
+	executor     ObjectDeleteExecutor
+	pollInterval time.Duration
+}
+
+// NewObjectDeleteWaiter is a constructor for ObjectDeleteExecutor.
+func NewObjectDeleteWaiter(c ObjectDeleteExecutor) ObjectDeleteWaiter {
+	return ObjectDeleteWaiter{executor: c, pollInterval: defaultPollInterval}
+}
+
+// SetPollInterval allows rewrite default poll interval.
+func (w *ObjectDeleteWaiter) SetPollInterval(interval time.Duration) {
+	w.pollInterval = interval
+}
+
+// ObjectDelete marks an object for deletion from the container using NeoFS API protocol.
+//
+// See details in [client.Client.ObjectDelete].
+func (w ObjectDeleteWaiter) ObjectDelete(ctx context.Context, containerID cid.ID, objectID oid.ID, prm client.PrmObjectDelete) (oid.ID, error) {
```

Comment:
do we really need it? what can that say about an object?

cc @roman-khimov @cthulhu-rider

Reply:
For instance: we may put object with the same name. It is useful if we want to have unique object name or another attribute in our system

Reply:
I'd say this has a similar set of guarantees to `ObjectPut` and we've never checked that this way. Names are just attributes and NeoFS can't (and shouldn't) guarantee their uniqueness now, so likely this doesn't add a lot of value.

Reply:
> we may put object with the same name

i meant that if _one_ node answers that it has removed an object it does not mean any other nodes have done it too. if it is the `Pool` that implements the nested interface, it is even possible (theoretically) to get an object from one node which (objects) is "removed" in another

Reply:
Ok, should I remove this waiter?

Reply:
I think so. Pool never had it, no known code does it. Can be added later if really needed.

Reply:
Removed